### PR TITLE
[Bugfix:TAGrading] Peer grading Total and Subtotals Scores

### LINKED ISSUE #5471: Totals & Subtotals while Grading (from TA & Peer View)
Just as a reminder...  this is what the rubric panel  looks like when a TA is grading a Autograde + Manual Grading (but no peer grading) assignment:  ( this shouldn't change )

autograding+tagrading

When a student/peer is grading (their grading is incomplete), this is what they should see:

partial_peer_grading

When the student/peer is finished grading, this is what they should see:
Note: the student/peer should never see any information about the autograding total, or the TA scores, or the other peer scores, or the overall Total.  
The peer can see the autograding panel with the autograding test cases, but we won't include the autograding score in the rubric panel. 

completed_peer_grading

This is what a TA should see for an assignment with both non-peer and peer Manual grading:
Note: normally TAs do not complete the peer components/marks.
If the assignment includes Autograding, those lines & scores are included as indicated.

no_ta_peer_grade

If the TA does enter their own scores for the peer components/marks, it will look like this:

ta_did_peer_grading

@emaicus @jchulton @josharsh    Comments/Questions?

Also tagging @mukul-kmr-jha  -- who is refactoring this webpage panel layout system

### PR DESCRIPTION
**Pull request by jchulton**
### Please check if the PR fulfills these requirements:

* [ ] The PR title and message follows our [guidelines](https://submitty.org/developer/how_to_contribute)
* [ ] Tests for the changes have been added/updated (if possible)
* [ ] Documentation has been updated/added if relevant

PR #5754 with bug fixes (remade due to original PR being inaccessible on a fork)

### What is the current behavior?
closes #5471  Currently, when a non-student grades peer and nonpeer components, only the total of that instructor's grading is shown: total peer and non-peer scores are not shown, peer graders also see the autograding results on their rubric.

### What is the new behavior?
When a non-student grades peer and non-peer components, there total peer and non-peer scores are shown separately. The total uses the combined peer score shown on the peer panel. Peers now only see the score they gave displayed on the rubric when grading.
On The details page for peer graders, autograding points and total points are now hidden.

### Other information?
Testing works best with "Closed Peer Team Homework" as it has autograding and is a peer assignment.
If the totals seem to not be appearing correctly do a hard reload

### CODE REVIEW

**Review by emaicus — CHANGES_REQUESTED** [review]
After installing this, pre-existing gradeables list manual grading to be `0/0` for both peers and instructors for already graded assignments. After grading something, the score updates to `N/0`, where `N` is the correct sum of points awarded.

Further, it doesn't look like this PR conforms to the new names for each of the point categories introduced in the associated issue.

**Comment by emaicus** [comment]
@jchulton, I had another look at this PR and encountered the totalling bug below. The steps that I took were to:
1. Log in as instructor and inspect the `Released Peer PDF Homework` gradeable
2. Log out of the instructor account
3. Log in as `aphacker` and inspect the `Released Peer PDF Homework` gradeable

wrong_total

**Comment by jchulton** [comment]
Fixed the bug above.

**Comment by shailpatels** [comment]
I don't think totals are being summed up correctly still for existing gradeables or does this only effect new gradeables?

image

**Review by shailpatels — APPROVED** [review]
Peer grading totals are calculated as defined by #5471

A combination of peer and normal components works fine and totals are calculated correctly, adding a peer component to an existing one also works correctly. Score breakdown is correct and peers do not see the total rubric information on the peer tab, TAs and instructors can see both.

*Note* The combined peer total badge only shows in the peer tab and not on the rubric tab, I don't think this a major problem and we can probably display the information in both places if needed later.

**Comment by bmcutler** [comment]
@loadedskiffer 
Please test this and figure out if it is finished, or if not, please finish whatever is missing/incorrect

File: site/app/controllers/grading/ElectronicGraderController.php [hunk 1]
```diff
@@ -199,6 +199,14 @@ private function setAllGradAllGrading($student_array) {
         }
         return $final_grading_info;
     }
+    /**
+     * Route for Getting Gradeable
+     * @Route("/courses/{_semester}/{_course}/gradeable/{gradeable_id}/getUserGroup", methods={"GET"})
+     */
+    public function getUserGroup($gradeable_id) {
```
Comment (williamjallen) [inline site/app/controllers/grading/ElectronicGraderController.php:206]: Can types be added for the parameter and return type?

File: site/app/libraries/database/DatabaseQueries.php [hunk 2]
```diff
@@ -1889,6 +1889,10 @@ public function getUnregisteredStudentsWithRotatingSection() {
         return $this->course_db->rows();
     }
 
+//    public function getComponentGrader() {
+//        $this->course_db->query("SELECT gcd_grader_id FROM gradeable_component_data WHERE")
+//    }
```
Comment (williamjallen) [inline site/app/libraries/database/DatabaseQueries.php:1894]: Does this need to be removed?

File: site/app/models/gradeable/GradedComponent.php [hunk 3]
```diff
@@ -32,6 +32,9 @@ class GradedComponent extends AbstractModel {
     /** @prop @var string Id of the component this grade is attached to */
     protected $component_id = 0;
 
+    /** If the component is peer */
+    private $is_peer = false;
```
Comment (williamjallen) [inline site/app/models/gradeable/GradedComponent.php:36]: Can a docstring be added here?

File: site/app/models/gradeable/GradedComponent.php [hunk 4]
```diff
@@ -118,6 +120,14 @@ public function toArray() {
     public function getComponent() {
         return $this->component;
     }
+    /**
+     * Gets if the component is peer
+     * @return bool
+     */
+    public function isPeer() {
```
Comment (williamjallen) [inline site/app/models/gradeable/GradedComponent.php:127]: perhaps add an inline return type here?

File: site/app/models/gradeable/GradedComponentContainer.php [hunk 5]
```diff
@@ -227,6 +227,59 @@ public function getTotalScore(User $grader = null) {
         return NumberUtils::roundPointValue($points_earned, $this->ta_graded_gradeable->getGradedGradeable()->getGradeable()->getPrecision());
     }
 
+    /**
+     * Gets the score the submitter received for this peer component, rounded
+     *  to the precision of the gradeable
+     * @return float
+     */
+    public function getTotalPeerScore(User $grader = null) {
```
Comment (williamjallen) [inline site/app/models/gradeable/GradedComponentContainer.php:235]: Perhaps declare an inline return type here too?

File: site/app/models/gradeable/GradedComponentContainer.php [hunk 6]
```diff
@@ -227,6 +227,59 @@ public function getTotalScore(User $grader = null) {
...
+
+        return $points_earned;
+    }
+
+    /**
+     * Gets the ta score the submitter received for this component, rounded
+     *  to the precision of the gradeable
+     * @return float
+     */
+    public function getTotalTaScore(User $grader = null) {
```
Comment (williamjallen) [inline site/app/models/gradeable/GradedComponentContainer.php:258]: same thing here to assist with static analysis

File: site/app/models/gradeable/GradedComponentContainer.php [hunk 7]
```diff
@@ -227,6 +227,59 @@ public function getTotalScore(User $grader = null) {
...
+                }
+            }
+            else {
+                $points_earned = 0;
+                $number_of_graders = 0;
+            }
+        }
+        // Note: this is called 'safeCalcPercent', but it does not clamp the output to 1.0
+        // Note: clamp count(...) to be at least 1 so safeCalcPercent doesn't return NaN
+        //$points_earned = Utils::safeCalcPercent($points_earned, max(1, $number_of_graders));
```
Comment (williamjallen) [inline site/app/models/gradeable/GradedComponentContainer.php:279]: Was this meant to be uncommented or should it be removed?

File: site/app/models/gradeable/TaGradedGradeable.php [hunk 8]
```diff
@@ -255,6 +255,30 @@ public function getTotalScore($grader = null) {
         }
         return $points_earned;
     }
+    /**
+     * Gets the instructor/ta score the student earned
+     * @return float
+     */
+    public function getTotalTaScore($grader = null) {
```
Comment (williamjallen) [inline site/app/models/gradeable/TaGradedGradeable.php:262]: same thing here about the return type

File: site/public/js/ta-grading-rubric.js [hunk 9]
```diff
@@ -1270,19 +1270,33 @@ function getGradedComponentFromDOM(component_id) {
...
 /**
  * Gets the scores data from the DOM (auto grading earned/possible and ta grading possible)
  * @return {Object}
  */
 function getScoresFromDOM() {
+    let gradeable_id = getGradeableId();
     let dataDOMElement = $('#gradeable-scores-id');
-
-    // Get the TA grading scores
+    var user_group_txt = $.ajax({
```
Comment (williamjallen) [inline site/public/js/ta-grading-rubric.js:1280]: Any reason why this (and lines 1291 and 1292) are marked `var` instead of `let`?

Comment (williamjallen) [reply site/public/js/ta-grading-rubric.js:1280]: Also, does this need custom functionality or would a standard JQuery GET request be simpler?

**Review by williamjallen — CHANGES_REQUESTED** [review]

File: site/app/libraries/database/DatabaseQueries.php (hunk 2 above)
Comment (loadedskiffer) [reply site/app/libraries/database/DatabaseQueries.php]: yes you're right

File: site/app/models/gradeable/GradedComponentContainer.php (hunk 7 above)
Comment (loadedskiffer) [reply site/app/models/gradeable/GradedComponentContainer.php]: removed

File: site/public/js/ta-grading-rubric.js (hunk 9 above)
Comment (loadedskiffer) [reply site/public/js/ta-grading-rubric.js]: I'm not sure how I would simplify it

Comment (williamjallen) [reply site/public/js/ta-grading-rubric.js]: Would [this](https://api.jquery.com/jquery.get/) serve the same purpose here/be simpler?

**Review by williamjallen — APPROVED** [review]

**Review by eddiekrystowski — APPROVED** [review]

**Comment by bmcutler** [comment]
@loadedskiffer 
Please resolve merge conflicts.  Thanks

**Review by bmcutler — CHANGES_REQUESTED** [review]
Unfortunately, this PR seems really broken.

When I edit the rubric to add a peer component to one of our grades released gradeables (it already has non peer components that are graded -- if that matters) it appears to save the new components & marks, but I get some pop up errors:

Screen Shot 2021-07-02 at 9 45 20 PM

Screen Shot 2021-07-02 at 9 45 02 PM

And then when I go to grade a student, and do the first grading of the peer component, I see this:

Screen Shot 2021-07-02 at 9 50 33 PM

The math is wrong and the categories are not what was requested in the issue:

Screen Shot 2021-07-02 at 9 54 35 PM

So I'm quite confused.  I think I've got the right branch installed...

**Comment by bmcutler** [comment]
So poking at this a little more, if there is no peer grading completed yet for the gradeable, it does not break out the peer grading subtotals:
notstarted


But then after I add the first peer score (and reload the browser tab):
started


This shouldn't happen, we should be able to tell if this gradeable has peer grading not by whether any peer scores are entered in the database, but by whether there is a component that is a peer component.

**Comment by loadedskiffer** [comment]
> So poking at this a little more, if there is no peer grading completed yet for the gradeable, it does not break out the peer grading subtotals:
> notstarted
> 
> But then after I add the first peer score:
> started
> 
> And if I reload the page then the peer panel updates too: (I also entered a grade from the TA account too)
> reload
> 
> This shouldn't happen, we should be able to tell if this gradeable has peer grading not by whether any peer scores are entered in the database, but by whether there is a component that is a peer component.

I fixed the pop up errors but as far as all of the other issues you've described, I have not been able to reproduce them. I used a gradeable with no peer components, added a peer component and everything still worked fine. Maybe try doing a ctrl-shift-r or inspect element and do an "empty cache and hard reload"

File: site/public/js/ta-grading-rubric.js [hunk 10]
```diff
@@ -96,6 +96,7 @@ function displayAjaxError(err) {
  * @return {Promise} Rejects except when the response returns status 'success'
  */
 function ajaxGetGradeableRubric(gradeable_id) {
+    console.log('here');
```
Comment (williamjallen) [inline site/public/js/ta-grading-rubric.js:99]: Should this be removed?

**Review by williamjallen — CHANGES_REQUESTED** [review]

**Review by williamjallen — APPROVED** [review]

**Review by bmcutler — APPROVED** [review]
Yay!  looks great!

**Pull request merged**